### ISSUE TITLE
Simplified / Automatic Lighting Generation

### ISSUE THREAD

**Issue by JoshPiper** [photonle/Photon#20]
## Feature
Be able to automatically generate simple patterns (side/side, alternating, incrementing, decrementing) for a set of colours and positions.

## Use Case
Simple components / legacy lights.

**Comment by JoshPiper** [comment]
## Implementation plan

### What has to be produced

Components hand-write two parallel tables:

- `COMPONENT.Sections[name]` — an array of *frames*; each frame is a list of `{ positionIndex, colour }` (`"_1"`/`"_2"` resolving through `DefaultColors`, or a literal like `"RED"` in legacy `EMV.Sections`).
- `COMPONENT.Patterns[name]` — named sequences of frame indices, `0` = blank.

So this feature is a pure table-generation problem: positions + colours + a generator name → frames + sequences. No runtime or rendering changes.

`fedsig_viper` shows the phase convention the generator should target:

```
code1  = { 1,1,1,0 }                -- unphased core
code1A = { 1,1,1,0, 0,0,0,0 }       -- core + blank half
code1B = { 0,0,0,0, 1,1,1,0 }       -- halves swapped
code2A = { 1,0,1,1,1, 0,2,2,2,0 }   -- two-colour core
code2B = { 0,2,2,2,0, 1,0,1,1,1 }   -- halves swapped
```

Every `B` is its `A` rotated by exactly half.

### Module

New `Photon.Patterns` in `lua/autorun/photon/shared/sh_patterns.lua`, included from `emv_init.lua` (`sh_` prefix ⇒ shared realm, loads before `library/emv_auto.lua`). Pure functions — no entity, realm or render dependencies.

```lua
Photon.Patterns.Generators[name] = function( lights, opts ) --> frames, sequence
Photon.Patterns.BuildSection( spec )                        --> frames, patterns
Photon.Patterns.Build( spec )                               --> sections, patterns
```

### Spec

```lua
COMPONENT.Generate = {
    ["auto_bar_main"] = {
        Positions = { 1, 2, 3, 4, 5, 6, 7, 8 },   -- defaults to all of COMPONENT.Positions
        Colors    = { "_1", "_2" },
        Layout    = "mirror",                      -- colour -> position mapping
        Order     = "centerout",                   -- traversal order for sweeps
        Phases    = true,                          -- emits <name>, <name>A, <name>B
        Patterns  = {
            code1   = { "flash",      Hold = 3 },
            code2   = { "sidetoside", Hold = 4, Split = "oddeven" },
            code3   = { "alternate",  Hold = 1, Gap = 1 },
            right   = { "increment",  Mode = "build" },
            left    = { "decrement",  Mode = "build" },
            diverge = { "increment",  Mode = "ramp", Window = 3 },
            chase   = { "increment",  Mode = "wrap", Window = 2 },
        },
    },
}
```

### Generators

Shared options: `Hold` (ticks per step), `Gap` (blank ticks between steps), `Repeat`.

| Name | Behaviour |
| --- | --- |
| `flash` | all lights on / off — the single-lighthead legacy case |
| `sidetoside` | split the lights and flash each half; `Split = "half"` (default), `"oddeven"`, or two explicit index lists |
| `alternate` | interleaved stride split, `Groups` (default 2), cycled |
| `increment` | steps a window along the light order (see below) |
| `decrement` | `increment` over the reversed order |

### Window modes

`increment`/`decrement` take a `Mode` and a `Window` (default 1):

| `Mode` | Behaviour | Steps |
| --- | --- | --- |
| `"build"` | window grows 1 → all lights and stays lit; the existing `left`/`right`/`diverge` sweeps on the TDM interior bar | `#lights` |
| `"ramp"` | fade in-out — grows 1 → `Window` as it enters, travels at full width, shrinks back to 1 as it exits; clipped at both ends, no wrap | `#lights + Window - 1` |
| `"wrap"` | continuous — always `Window` wide; running off the end splits the tail and re-enters it at the start | `#lights` |

At `Window = 1`, `ramp` and `wrap` produce the same single travelling light.

### Colour and position layout

Two orthogonal fields, because "which colour" and "in what order" are separate questions.

`Layout` — colour per position:

| Value | 6 positions, `Colors = { "_1", "_2" }` |
| --- | --- |
| `"cycle"` (default) | `1 2 1 2 1 2` — interleaved |
| `"pairs"` | `1 1 2 2 1 1` — paired round-robin |
| `"blocks"` | `1 1 1 2 2 2` — contiguous split, left to right |
| `"mirror"` | `1 2 2 2 2 1` — split about the centre, worked outward |
| array | `{ "_1", "_2", "_2", "_1", "_3", "_3" }` — per-position, wins over everything |
| function | `function( slot, positionIndex, count ) return colour end` |

`Order` — traversal order for sweeps:

| Value | Positions 1–6 |
| --- | --- |
| `"forward"` (default) | `1 2 3 4 5 6` |
| `"reverse"` | `6 5 4 3 2 1` |
| `"centerout"` | `{3,4} {2,5} {1,6}` — the diverge sweep, mirrored pairs stepping together |
| `"outsidein"` | `{1,6} {2,5} {3,4}` |
| array | `{ 5, 6, 4, 3 }`, or nested `{ {5,6}, {4,7}, {3,8} }` to step several positions per frame |

The nested-array form is the general escape hatch; `centerout` and `outsidein` are sugar over it. Both fields are per-section, so a bar with a centre-out takedown row and a left-to-right corner strip declares two sections — which is how `Sections` already partitions lights.

### Phases

`Phases = true` emits three names per pattern from one core sequence:

- `<name>` — the core alone, as the no-phase fallback. This has to exist: `EMVU` resolves patterns as `patternIndex .. ( autoData.Phase or "" )`, so an unphased component asks for the bare name.
- `<name>A` — core padded to a double-length cycle.
- `<name>B` — the same cycle with its halves swapped.

`Phases = { "A", "B", "C" }` splits into that many equal slots. Explicit rotations (`Phases = { A = 0, B = 0.5 }`) stay available for odd cases.

### Other behaviour

- **Frame interning** — `Build` keys frames canonically so patterns sharing a section reuse frame indices instead of duplicating them.
- **Colour cycling** — `ColorCycle = { "_1", "_2" }` repeats the generated sequence once per colour, which is how `code2A`/`code2B` above get their two-colour cores.
- **Validation** — out-of-range position index, unknown generator name, empty light set and duplicate section names each `PhotonError` with the component and section name, matching the warn-and-continue style already in `AddAutoComponent`.

### Integration

`EMVU:AddAutoComponent` expands `component.Generate` into `Sections`/`Patterns` *before* the `Base`/`table.Inherit` branch, so inherited components pick up the expanded tables. Generated and hand-written section names must be disjoint — a collision errors rather than silently merging two array shapes. `Modes` stays hand-written; generated patterns are just names it can reference.

Legacy vehicle files get this for free: `EMV.Sections`/`EMV.Patterns` use the identical frame shape, so they can call `Photon.Patterns.BuildSection` directly with literal colour names.

### Files

- **new** `lua/autorun/photon/shared/sh_patterns.lua` — the module, LDoc-annotated (`@module Photon.Patterns`)
- **edit** `lua/autorun/emv_init.lua` — `Photon.include( "shared/sh_patterns.lua" )`
- **edit** `lua/autorun/photon/library/emv_auto.lua` — expand `COMPONENT.Generate`
- **new** `lua/tests/photon/patterns.lua` — GLuaTest group

### Testing

GLuaTest already runs on PRs, and these are pure table functions:

- light resolution: explicit list, defaults, every named `Layout` and `Order`, plus the array and function escape hatches
- each generator's frames and sequence, including `Hold`/`Gap`
- each window mode's step count and the `wrap` tail-split; `Window = 1` equivalence between `ramp` and `wrap`
- phase identity: `B` is `A` rotated by half, for any generated pattern
- frame interning and colour cycling
- error paths: bad position index, unknown generator, section name collision
- **regression anchor:** regenerate `fedsig_viper`'s `code1`/`code1A`/`code1B` and `code2A`/`code2B` and assert equality with the literal tables in the shipped file

Not `whelen_600` — it predates this convention (its bare `code1` is already a padded A-phase and its `code1A` is really the B), so matching it byte-for-byte would mean encoding the older layout.

### Out of scope / risks

- No shipped component gets converted to `Generate` in this PR. `componentlint.lua` expects literal `Sections`/`Patterns` and is currently manual-only pending an update, so the feature stays purely additive; conversions can follow once the linter is updated.
- Docs need no manual work — the Documentation workflow regenerates LDoc on push to `development`.
- No in-game verification in CI; a component author should smoke-test one generated component before release.
- Related: #21 (component templates) would naturally consume this once it lands.

### Steps

1. `Photon.Patterns` module — light resolution, `Layout`/`Order`, frame interning, `Build`/`BuildSection`
2. The five generators, window modes, phase pairing, colour cycling
3. `COMPONENT.Generate` expansion in `AddAutoComponent`, with validation
4. GLuaTest group including the `fedsig_viper` regression anchor
5. LDoc pass

One PR against `development`.
